adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1395 @@
+		device.Oaid = deviceId.Oaid[0]
+	}
+	if len(deviceId.Gaid) > 0 {
+		device.Gaid = deviceId.Gaid[0]
+	}
+	if len(deviceId.Imei) > 0 {
+		device.Imei = deviceId.Imei[0]
+	}
+	if len(deviceId.ClientTime) > 0 {
+		device.ClientTime = getClientTime(deviceId.ClientTime[0])
+	}
+	// IsTrackingEnabled = 1 - DNT
+	if openRTBRequest.Device != nil && openRTBRequest.Device.DNT != nil {
+		if device.Oaid != "" {
+			device.IsTrackingEnabled = strconv.Itoa(1 - int(*openRTBRequest.Device.DNT))
+		}
+		if device.Gaid != "" {
+			device.GaidTrackingEnabled = strconv.Itoa(1 - int(*openRTBRequest.Device.DNT))
+		}
+	}
+	return nil
+}
+
+// get network information for HuaweiAds request
+func getHuaweiAdsReqNetWorkInfo(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	var network Network
+	if openRTBRequest.Device != nil {
+		if openRTBRequest.Device.ConnectionType != nil {
+			network.Type = int32(*openRTBRequest.Device.ConnectionType)
+		} else {
+			network.Type = DefaultUnknownNetworkType
+		}
+
+		var cellInfos []CellInfo
+		if openRTBRequest.Device.MCCMNC != "" {
+			var arr = strings.Split(openRTBRequest.Device.MCCMNC, "-")
+			network.Carrier = 0
+			if len(arr) >= 2 {
+				cellInfos = append(cellInfos, CellInfo{
+					Mcc: arr[0],
+					Mnc: arr[1],
+				})
+				var str = arr[0] + arr[1]
+				if str == "46000" || str == "46002" || str == "46007" {
+					network.Carrier = 2
+				} else if str == "46001" || str == "46006" {
+					network.Carrier = 1
+				} else if str == "46003" || str == "46005" || str == "46011" {
+					network.Carrier = 3
+				} else {
+					network.Carrier = 99
+				}
+			}
+		}
+		network.CellInfo = cellInfos
+	}
+	request.Network = network
+}
+
+// get regs information for HuaweiAds request
+func getHuaweiAdsReqRegsInfo(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	var regs Regs
+	if openRTBRequest.Regs != nil {
+		regs.Coppa = int32(openRTBRequest.Regs.COPPA)
+	}
+	request.Regs = regs
+}
+
+// get geo information for HuaweiAds request
+func getHuaweiAdsReqGeoInfo(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	var geo Geo
+	if openRTBRequest.Device != nil && openRTBRequest.Device.Geo != nil {
+		geo.Lon = float32(openRTBRequest.Device.Geo.Lon)
+		geo.Lat = float32(openRTBRequest.Device.Geo.Lat)
+		geo.Accuracy = int32(openRTBRequest.Device.Geo.Accuracy)
+		geo.Lastfix = int32(openRTBRequest.Device.Geo.LastFix)
+	}
+	request.Geo = geo
+}
+
+func unmarshalExtImpHuaweiAds(openRTBImp *openrtb2.Imp) (*openrtb_ext.ExtImpHuaweiAds, error) {
+	var bidderExt adapters.ExtImpBidder
+	var huaweiAdsImpExt openrtb_ext.ExtImpHuaweiAds
+	if err := json.Unmarshal(openRTBImp.Ext, &bidderExt); err != nil {
+		return nil, errors.New("Unmarshal: openRTBImp.Ext -> bidderExt failed")
+	}
+	if err := json.Unmarshal(bidderExt.Bidder, &huaweiAdsImpExt); err != nil {
+		return nil, errors.New("Unmarshal: bidderExt.Bidder -> huaweiAdsImpExt failed")
+	}
+	if huaweiAdsImpExt.SlotId == "" {
```

Comment:
Since these are required in your JSON schema, the entire prebid request will fail with an error (no bidders will bid) if any of these parameters are missing. I have no problem with verifying this again here, but thought you should know.

Reply:
yes, I know, all of them are required parameters ，and I confirm.